File: 0057-insert-interval/0057-insert-interval.cpp
class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        int n = intervals.size();

        // find the first interval with interval[1] >= newInterval[0]
        int left = -1, right = n;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (intervals[mid][1] >= newInterval[0]) {
                right = mid;
            } else {
                left = mid;
            }
        }
        int firstIdx = right;

        // find the last interval with interval[0] <= newInterval[1]
        left = -1;
        right = n;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;
            if (intervals[mid][0] <= newInterval[1]) {
                left = mid;
            } else {
                right = mid;
            }
        }
        int lastIdx = left;

        vector<vector<int>> ans;
        if (lastIdx == -1) {
            ans.push_back(newInterval);
            for (auto& iv: intervals) ans.push_back(iv);
        } else if (firstIdx == n) {
            for (auto& iv: intervals) ans.push_back(iv);
            ans.push_back(newInterval);
        } else if (lastIdx < firstIdx) {
            for (int i = 0; i <= lastIdx; i++) ans.push_back(intervals[i]);
            ans.push_back(newInterval);
            for (int i = firstIdx; i < n; i++) ans.push_back(intervals[i]);
        } else {
            int minStart = min(intervals[firstIdx][0], newInterval[0]);
            int maxEnd = max(intervals[lastIdx][1], newInterval[1]);
            for (int i = 0; i < n; i++) {
                if (i < firstIdx || i > lastIdx) {
                    ans.push_back(intervals[i]);
                } else if (i == firstIdx) {
                    ans.push_back({ minStart, maxEnd });
                }
            }
        }
        return ans;
    }

    //             -- ---- ----- ----
    //  --------


    // Approach: handle 3 cases
    // - non-overlapping intervals at the left of newInterval
    // - overlapping intervals with newInterval
    // - non-overlapping intervals at the right of newInterval

    // Time: O(n)
    // Space: O(1)

    // vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
    //     int n = intervals.size();
    //     int i = 0;

    //     if (n == 0) return { newInterval };

    //     vector<vector<int>> ans;

    //     // 1st case
    //     while (i < n && intervals[i][1] < newInterval[0]) {
    //         ans.push_back(intervals[i]);
    //         i++;
    //     }

    //     // 2nd case
    //     int min_left = newInterval[0];
    //     int max_right = newInterval[1];
    //     while (i < n && intervals[i][0] <= newInterval[1]) {
    //         min_left = min(min_left, intervals[i][0]);
    //         max_right = max(max_right, intervals[i][1]);
    //         i++;
    //     }
    //     ans.push_back({ min_left, max_right });

    //     // 3rd case
    //     while (i < n) {
    //         ans.push_back(intervals[i]);
    //         i++;
    //     }

    //     return ans;
    // }
};

File: 0057-insert-interval/0057-insert-interval.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        n = len(intervals)
        
        if n == 0:
            return [newInterval]
        
        ans = []
        idx = 0

        while idx < n and intervals[idx][1] < newInterval[0]:
            ans.append(intervals[idx])
            idx += 1

        merged_min = newInterval[0]
        merged_max = newInterval[1]
        while idx < n and intervals[idx][0] <= newInterval[1]:            
            merged_min = min(merged_min, intervals[idx][0])
            merged_max = max(merged_max, intervals[idx][1])
            idx += 1
        ans.append([merged_min, merged_max])

        while idx < n:
            ans.append(intervals[idx])
            idx += 1

        return ans

"""
test case: intervals = [[1,3]], newInterval = [4,4]
trace:
idx = 0
ans = [[1,3]]
idx = 1
merged_min = 4, merged_max = 4
ans = [[1,3], [4,4]]

test case: intervals = [[1,3]], newInterval = [0,0]
trace:
idx = 0
merged_min = 0, merged_max = 0
ans = [[0,0]]
ans = [[0,0], [1,3]]

test case: intervals = [[1,3], [6,9]], newInterval = [2,5]
trace:
idx = 0
ans = []
merged_min = 2, merged_max = 5
merged_min = 1, merged_max = 5
idx = 1
ans = [[1,5]]
idx = 2
ans = [[1,5], [6,9]]
"""